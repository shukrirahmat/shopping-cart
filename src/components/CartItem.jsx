import styles from "../styles/CartItem.module.css";
import PropTypes, { func } from "prop-types";

function CartItem({ item , removeItem, modifyAmount}) {

  function pricetoString(price) {
    return price.toFixed(2) + "$";
  }

  function removeClick() {
    removeItem(item.id);
  }

  function handleChange(e) {
    const value = e.target.value.replace(/[^0-9]/g, "");
    let num;
    if (!value) num = 0;
    else num = parseInt(value);
    if (num !== item.amount) {
        modifyAmount(item.id, num);
    }
  }

  function handleAdd() {
    modifyAmount(item.id, item.amount + 1);
  }

  function handleReduce() {
    if (item.amount > 1) modifyAmount(item.id, item.amount - 1);
  }

  return (
    <div className={styles.container}>
      <p className={styles.title}>{item.title}</p>
      <div className={styles.imageContainer}>
        <img src={item.image}></img>
      </div>
      <div className={styles.amount}>
        <button onClick={handleReduce}>-</button>
        <input type="text" value={item.amount} onChange={handleChange}></input>
        <button onClick={handleAdd}>+</button>
      </div>
      <div className={styles.priceContainer}>
        <p>Price: {pricetoString(item.price)}</p>
        <p>Total: {pricetoString(item.price * item.amount)}</p>
      </div>
      <button className={styles.removeButton} onClick={removeClick}>Remove</button>
    </div>
  );
}

CartItem.propTypes = {
  item: PropTypes.object,
  removeItem: PropTypes.func,
  modifyAmount: PropTypes.func
};

export default CartItem;
