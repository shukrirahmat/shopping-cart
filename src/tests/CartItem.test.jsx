import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CartItem from "../components/CartItem";

describe("cart item component", () => {
  const item = {
    id: "69",
    title: "test product",
    price: 13.5,
    description: "lorem ipsum set",
    image: "https://i.pravatar.cc",
    category: "electronic",
    amount: 3,
  };

  it("renders cart item", () => {
    const { container } = render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={() => {}}
        modifyAmount={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("should call the modify function when number is entered", async () => {
    const modifyAmount = vi.fn();
    const user = userEvent.setup();
    render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={() => {}}
        modifyAmount={modifyAmount}
      />
    );
    const input = screen.getByRole("textbox");

    await user.type(input, "a2b");

    expect(modifyAmount).toHaveBeenCalled();
  });

  it("should not call the modify fuction when non numeric entered", async () => {
    const modifyAmount = vi.fn();
    const user = userEvent.setup();
    render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={() => {}}
        modifyAmount={modifyAmount}
      />
    );
    expect(modifyAmount).not.toHaveBeenCalled();

    const input = screen.getByRole("textbox");
    await user.type(input, "abc");

    expect(modifyAmount).not.toHaveBeenCalled();
  });

  it("should call the modify function when exiting focus with less than 1", async () => {
    const modifyAmount = vi.fn();
    const user = userEvent.setup();
    render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={() => {}}
        modifyAmount={modifyAmount}
      />
    );
    const input = screen.getByRole("textbox");

    await user.clear(input);
    await user.tab()

    expect(modifyAmount).toHaveBeenCalled();
  });

  it("should not call the modify function when exiting focus with more than 1", async () => {
    const modifyAmount = vi.fn();
    const user = userEvent.setup();
    render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={() => {}}
        modifyAmount={modifyAmount}
      />
    );
    const input = screen.getByRole("textbox");

    await user.click(input);
    await user.tab()

    expect(modifyAmount).not.toHaveBeenCalled();
  });

  it("should call the modify function when + button is clicked", async () => {
    const modifyAmount = vi.fn();
    const user = userEvent.setup();
    render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={() => {}}
        modifyAmount={modifyAmount}
      />
    );
    const button = screen.getByRole("button", {name: "+"});

    await user.click(button)

    expect(modifyAmount).toHaveBeenCalled();
  });

  it("should call the modify function when - button is clicked", async () => {
    const modifyAmount = vi.fn();
    const user = userEvent.setup();
    render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={() => {}}
        modifyAmount={modifyAmount}
      />
    );
    const button = screen.getByRole("button", {name: "-"});

    await user.click(button)

    expect(modifyAmount).toHaveBeenCalled();
  });

  it("should not call the modify function when - button is clicked at 1", async () => {
    const modifyAmount = vi.fn();
    const user = userEvent.setup();
    render(
      <CartItem
        item={{
            id: "69",
            title: "test product",
            price: 13.5,
            description: "lorem ipsum set",
            image: "https://i.pravatar.cc",
            category: "electronic",
            amount: 1,
          }}
        key={item.id}
        removeItem={() => {}}
        modifyAmount={modifyAmount}
      />
    );
    const button = screen.getByRole("button", {name: "-"});

    await user.click(button)
    await user.click(button)

    expect(modifyAmount).not.toHaveBeenCalled();
  });

  it("should not call the modify function when no button is clicked", async () => {
    const modifyAmount = vi.fn();
    render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={() => {}}
        modifyAmount={modifyAmount}
      />
    );

    expect(modifyAmount).not.toHaveBeenCalled();
  });

  it("should call the remove function when - remove button is clicked", async () => {
    const removeItem = vi.fn();
    const user = userEvent.setup();
    render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={removeItem}
        modifyAmount={() => {}}
      />
    );
    const button = screen.getByRole("button", {name: /Remove/i});

    await user.click(button)

    expect(removeItem).toHaveBeenCalled();
  });

  it("should not call the remove function when - remove button is not clicked", async () => {
    const removeItem = vi.fn();
    render(
      <CartItem
        item={item}
        key={item.id}
        removeItem={removeItem}
        modifyAmount={() => {}}
      />
    );

    expect(removeItem).not.toHaveBeenCalled();
  });
});
