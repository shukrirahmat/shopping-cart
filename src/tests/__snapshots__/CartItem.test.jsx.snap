// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cart item component > renders cart item 1`] = `
<div>
  <div
    class="_container_0d2c86"
  >
    <p
      class="_title_0d2c86"
    >
      test product
    </p>
    <div
      class="_imageContainer_0d2c86"
    >
      <img
        src="https://i.pravatar.cc"
      />
    </div>
    <div
      class="_amount_0d2c86"
    >
      <button>
        -
      </button>
      <input
        type="text"
        value="3"
      />
      <button>
        +
      </button>
    </div>
    <div
      class="_priceContainer_0d2c86"
    >
      <p>
        Price: 
        13.50$
      </p>
      <p>
        Total: 
        40.50$
      </p>
    </div>
    <button
      class="_removeButton_0d2c86"
    >
      Remove
    </button>
  </div>
</div>
`;
